### ISSUE TITLE
Model Position Incorrectly Changed from Centre for some Coordinates

### ISSUE THREAD

**Issue by programmer443** [mapbox/mapbox-maps-ios#2176]
## Environment

- Xcode version: 15.3
- iOS version: 17.4
- Devices affected:
- Maps SDK Version: 11.1.0

## Observed behavior and steps to reproduce

I encountered an issue where the position of a model in my Mapbox implementation is being inaccurately adjusted for certain coordinates but camera position is fine. The model, which should be centered at specified coordinates, is shifting its position unexpectedly. This behavior occurs despite correct implementation of adding the model, setting camera parameters.

**Code for adding model**

```

let plane = Bundle.main.url(
    forResource: MapViewModel.MapViewLabels.modelName,
    withExtension: MapViewModel.MapViewLabels.modelExt)!.absoluteString
var source = GeoJSONSource(id: MapViewLabels.sourceID)
source.maxzoom = 4
try? mapProxy?.map!.addStyleModel(modelId: MapViewLabels.modelID, modelUri: plane)
var planeFeature = Feature(geometry: Point(CLLocationCoordinate2D(latitude: self.latitude, longitude: self.longitude)))
planeFeature.properties = [MapViewLabels.modelKey: .string(MapViewLabels.modelID)]
source.data = .featureCollection(FeatureCollection(features: [planeFeature]))
try? mapProxy?.map!.addSource(source)
var layer = ModelLayer(id: MapViewLabels.layerId, source: MapViewLabels.sourceID)
layer.modelId = .expression(Exp(.get) { MapViewLabels.modelKey })
layer.modelType = .constant(.common3d)
layer.modelScale = .constant(MapViewModel.MapViewLabels.modelScale)
layer.modelTranslation = .constant([0, 0, self.altitude.feetToMeter()])
let headingAngle = self.heading - 90
layer.modelRotation = .constant([0, 0, headingAngle])
layer.maxZoom = 23
layer.minZoom = 5
layer.modelRotationTransition = StyleTransition(duration: 0.1, delay: 0.1)
layer.modelTranslationTransition = StyleTransition(duration: 0, delay: 0)
layer.modelCutoffFadeRange = .constant(0.0)
try? mapProxy?.map!.addLayer(layer)
```

**Code for setting camera:**

```

let freeCam = self.mapProxy?.map?.freeCameraOptions          
freeCam?.altitude = alt
freeCam?.location = CLLocationCoordinate2D(latitude: lat, longitude: lng)
freeCam?.setPitchBearingForPitch(pitch, bearing: -head)
mapProxy?.map?.freeCameraOptions = freeCam!
```

## Expected behavior

Model should appear at the centre like camera 

**Picture**

![Model images slightly left from centre](https://github.com/mapbox/mapbox-maps-ios/assets/82497041/d3fdf041-4a83-4c12-a9f8-6b01b1f113d4)

**Comment by pjleonard37** [comment]
Hi @programmer443 - 

Thanks for sharing this report and the code sample. To help us replicate, could you provide a smaller example with actual values for the variables in your code for latitude, longitude, etc? This will help to isolate if the issue is in the SDK rather than with incorrect data. 

It would be especially helpful if you could recreate the issue outside of your application in a simple implementation of the SDK. For example, you could modify one of our [examples](https://github.com/mapbox/mapbox-maps-ios/tree/main/Apps/Examples/Examples).

**Comment by programmer443** [comment]
Hi @pjleonard37,

Thanks for your reply🙂. I updated my code with actual values for your understanding. Now, I hope so it will help you to reproduce the issue.

**Comment by pjleonard37** [comment]
Hi @programmer443 -- 

Thanks for sharing this sample project; I was able to replicate the behavior in your project. I think the issue is related to how you've set up the SwiftUI <--> UIKit code. I was able to strip down the code a bit more and have made some changes to center the plane model:

![Simulator Screenshot - iPhone 15 Pro - 2024-05-10 at 17 11 58](https://github.com/mapbox/mapbox-maps-ios/assets/7976026/42917322-7e28-44f9-a76a-7368f53fa54a)

Instead of using a Swift UI and UIKit together, I just implemented the solution in Swift UI, which resolves the misalignment you were seeing. Note that this code takes advantage of new SDK features released in the `v11.4.0-beta.1` pre-release. You will either need to use this pre-release or `main` to use these new features until the stable version of `v11.4.0` is released (in a few weeks). The key changes are our new tools for [Declarative Map Styling](https://docs.tilestream.net/ios/maps/api/latest/documentation/mapboxmaps/declarative-map-styling/), which should dramatically simplify your SwiftUI implementation. 

```
struct ExampleMapModel: View {
    let plane = Bundle.main.url(
        forResource: "A320",
        withExtension: "glb")
    var latitude: CGFloat = 42.97330
    var longitude: CGFloat = -75.085930

    var body: some View {
        MapReader { proxy in
            Map() {
                GeoJSONSource(id: "source-id")
                    .data(.featureCollection(planeFeatures))
                Model(id: "model-id-plane", uri: plane)
                ModelLayer(id: "model-layer-id", source: "source-id")
                    .modelId(Exp(.get) { "model-id-key" })
                    .modelType(.common3d)
                    .modelScale(x: 1, y: 1, z: 1)
                    .modelTranslation(x: 0, y: 0, z: 9753.60)
                    .modelRotation(x: 0, y: 0, z: 90)
                    .maxZoom(23)
                    .minZoom(5)
                    .modelCutoffFadeRange(0.0)
            }
            .mapStyle(.satelliteStreets)
            .onMapLoaded { _ in
                guard let map = proxy.map else { return }
                let freeCam = map.freeCameraOptions
                freeCam.altitude = 10000
                freeCam.location = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
                freeCam.setPitchBearingForPitch(0, bearing: 180.0)
                map.freeCameraOptions = freeCam
            }

        }
    }
}

var planeFeatures: FeatureCollection {
    var planeFeature = Feature(geometry: Point(CLLocationCoordinate2D(latitude: 42.97330, longitude: -75.085930)))
    planeFeature.properties = ["model-id-key": .string("model-id-plane")]
    return FeatureCollection(features: [planeFeature])
}
```

These features are currently experimental, so breaking changes may occur across versions. As we refine these features we are very interested in hearing your feedback! 

One other note: I'd recommend rotating the access token in your shared project. Instructions can be found here: https://docs.mapbox.com/accounts/guides/tokens/#rotating-access-tokens.

**Comment by programmer443** [comment]
@pjleonard37 just checking in on the status of this issue—any updates or ETA for a resolution?
cc: @mapbox-maps-ios

**Comment by pjleonard37** [comment]
Hi @programmer443 -- 

I have not been able to reproduce the issue you are seeing locally, and am still investigating. 

As a friendly reminder, Model layers are an experimental feature.

**Comment by programmer443** [comment]
Hi @pjleonard37 @astojilj,
Thank you for your assistance with my previous issue. The solution using SwiftUI has helped simplify the implementation significantly. However, it hasn't solved the actual issue I was experiencing.
The problem where the model wasn't centered correctly was due to the maxZoom setting in the GeoJSONSource. Without setting maxZoom, the model disappears randomly.

Related link: https://github.com/mapbox/mapbox-maps-ios/issues/2152

**Comment by pjleonard37** [comment]
Hi @programmer443 - 

I'm glad to hear that your implementation has been simplified using SwiftUI. 

> The problem where the model wasn't centered correctly was due to the maxZoom setting in the GeoJSONSource. Without setting maxZoom, the model disappears randomly.

Just to confirm, when you set the maxZoom the issue was resolved for you?

**Comment by programmer443** [comment]
Yes, if i set maxzoom to 6+ the model appears in center but it disappears randomly. If i set maxzoom below 6 the model doesn't disappear anymore but it's location shifts away from center.

**Comment by astojilj** [comment]
@programmer443 

Is maxzoom 6 used in this video? Could you please also set source's "buffer" to 512?

https://github.com/mapbox/mapbox-maps-ios/issues/2176#issuecomment-2129394486

**Comment by qandeelcc** [comment]
@astojilj there was no max zoom in the above video, and we didn't specifically change the source's buffer. Later when we realized that the model was disappearing due to missing maxZoom, we set its value to 8. With zoom value 8, the model was positioned in centered but it would still disappear in certain cases, specially when changing pitch. Will setting source's buffer fix the issue?

**Comment by qandeelcc** [comment]
@astojilj @pjleonard37 We've tried all suggested solutions but unfortunately couldn't make it work. Do you have any other ideas, or should we wait for the ModelLayer to get stable?

FYI, we are animating the model layer over a long distance using modelTranslation while keeping its coordinates the same. Do you think it could be the cause of model disappearance after some time?
